packages/concerto-tools/lib/codegen/fromcto/protobuf/protobufvisitor.js

```diff
@@ -0,0 +1,368 @@
+    }
+
+    /**
+     * Visitor design pattern
+     * @param {ClassDeclaration} classDeclaration - the object being visited
+     * @param {Object} parameters - the parameter
+     * @return {Object} the result of visiting or null
+     * @private
+     */
+    visitClassDeclaration(classDeclaration, parameters) {
+        debug('entering visitClassDeclaration', classDeclaration.getName());
+        return this.visitClassDeclarationCommon(classDeclaration, parameters);
+    }
+
+    /**
+     * Visit a Concerto class
+     * @param {ClassDeclaration} classDeclaration - the Concerto class being visited
+     * @param {Object} parameters - the parameters
+     * @private
+     */
+    visitClassDeclarationCommon(classDeclaration, parameters) {
+        debug('entering visitClassDeclarationCommon', classDeclaration.getName());
+
+        // Don't write abstract classes.
+        if (!classDeclaration.ast?.isAbstract) {
+            // Check if the class contains properties.
+            const classContainsProperties = classDeclaration.getProperties()?.length > 0;
+
+            // Write the beginning of the message and the opening bracket (the closing one as well if the class is empty).
+            parameters.fileWriter.writeLine(0, `message ${classDeclaration.getName()} {${classContainsProperties ? '' : '}\n'}`);
+
+            if (classContainsProperties) {
+                classDeclaration.getProperties()
+                    .map(
+                        (property) => {
+                            if (
+                                // If property starts with a $ and is not a $identifier.
+                                (
+                                    property.getName().charAt(0) === '$' &&
+                                    property.getName() !== '$identifier'
+                                ) ||
+                                // Or if property is an $identifier, but there already is another identifier defined.
+                                (
+                                    property.getName() === '$identifier' &&
+                                    typeof classDeclaration.getIdentifierFieldName() === 'string'
+                                )
+                            ) {
+                                // Then don't include this property.
+                                return;
+                            }
+                            // All the other properties should be added to the list of properties to be written in the file.
+                            return property.accept(this, parameters);
+                        }
+                    )
+                    // Make sure that we don't have any unidentified fields.
+                    .filter(proto3FieldObject => proto3FieldObject !== undefined)
+                    // Write the fields, adding a Proto3 index to them.
+                    .forEach(
+                        ({ preposition, type, fieldName }, i) => {
```

Comment:
I think the use of `I` here may be a problem, as it means that if the user reorders their list of properties in the CTO file (a non-breaking change) [or in a super-type CTO file] then the indexes used for the protobuf will all change (a breaking change for protobuf). Instead, I suggest that we sort the properties by name and use the sorted list to compute the protobuf index.